### PR TITLE
Add a reader for insat 3d imager data (`insat3d_img_l1b_h5`)

### PR DESCRIPTION
**Pull request by mraspaud**
This PR adds a reader for insat 3d L1B imager data in hdf5 format.

 - [x] Tests added 
 - [ ] Fully documented

### CODE REVIEW

**Comment by mraspaud** [comment]
The VIS channel
VIS3D

**Comment by simonrp84** [comment]
This PR currently requires the `xarray-datatree` library, which is a bit unintuitive to find. Would be good to have a note somewhere in the docs for this reader saying which library is required :)

**Comment by simonrp84** [comment]
`Is this any different from the other readers who have non-"standard" dependencies?`

Yes, I think so - the library needed here isn't so intuitive to find. The error message given by satpy is a generic reader failure message and the `debug_on` logs say that the `datatree` library is missing. I had to ask Martin where to find it as there's no datatree library on conda :)

**Comment by mraspaud** [comment]
Some composites created with this data:
- Cloud phase distinction 
cloud_phase_distinction_20221101_040009

- green snow 
green_snow_20221101_040009

- night fog (night microphysics)
night_fog_20221101_040009

**Comment by mraspaud** [comment]
pre-commit.ci autofix

**Comment by mraspaud** [comment]
pre-commit.ci autofix

File: satpy/etc/readers/insat3d_img_l1b_h5.yaml [hunk 1]
```diff
@@ -0,0 +1,162 @@
+reader:
+  name: insat3d_img_l1b_h5
```
Comment (djhoese) [inline satpy/etc/readers/insat3d_img_l1b_h5.yaml:2]: Is there a version of this L1B data that isn't HDF5?

File: satpy/etc/readers/insat3d_img_l1b_h5.yaml [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    name: VIS
+    resolution: 1000
+    wavelength: [0.525, 0.65, 0.775]
+    calibration:
+      reflectance:
+        standard_name: toa_bidirectional_reflectance
+        units: "%"
+      radiance:
+        standard_name: toa_outgoing_radiance_per_unit_wavelength
+        units: mW.cm-2.sr-1.micron-1
```
Comment (djhoese) [inline satpy/etc/readers/insat3d_img_l1b_h5.yaml:33]: Not sure we standardized on it, but are spaces and periods both accepted as udunits/CF compliant units?

File: satpy/readers/insat3d_img_l1b_h5.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    if resolution == 1000:
+        lonlat_suffix = "_VIS"
+    elif resolution == 8000:
+        lonlat_suffix = "_WV"
+    else:
+        lonlat_suffix = ""
+    return lonlat_suffix
+
+
+@lru_cache
```
Comment (djhoese) [inline satpy/readers/insat3d_img_l1b_h5.py:57]: Is this caching needed? This is only called from `open_datatree` and that function is only called from a cached property. If each resolution has its own file then is it buying us anything to cache?

File: satpy/readers/insat3d_img_l1b_h5.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+def open_dataset(filename, resolution=1000):
+    """Open a dataset for a given resolution."""
+    h5ds = xr.open_dataset(filename, engine="h5netcdf", chunks="auto")
+    h5ds_raw = xr.open_dataset(filename, engine="h5netcdf", chunks="auto", mask_and_scale=False)
+    ds = xr.Dataset()
+    ds.attrs = h5ds.attrs
+    if resolution in [1000, 4000, 8000]:
```
Comment (djhoese) [inline satpy/readers/insat3d_img_l1b_h5.py:64]: Flip this condition and raise the ValueError at the top of the method before you open the dataset at all.

File: satpy/readers/insat3d_img_l1b_h5.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            with suppress(ValueError):
+                ds = ds.rename({y_dim: "y"})
+
+        for lons in ["Longitude_VIS", "Longitude_WV"]:
+            with suppress(ValueError):
+                ds = ds.rename({lons: "Longitude"})
+
+        for lats in ["Latitude_VIS", "Latitude_WV"]:
+            with suppress(ValueError):
+                ds = ds.rename({lats: "Latitude"})
```
Comment (djhoese) [inline satpy/readers/insat3d_img_l1b_h5.py:95]: Maybe put these in one or functions to cleanup the code a bit (codescene doesn't like the large complexity)

**Review by djhoese — CHANGES_REQUESTED** [review]

File: satpy/etc/readers/insat3d_img_l1b_h5.yaml (hunk 1 above)
Comment (mraspaud) [reply satpy/etc/readers/insat3d_img_l1b_h5.yaml:2]: Yes, geotiff also according to the website

File: satpy/etc/readers/insat3d_img_l1b_h5.yaml (hunk 2 above)
Comment (mraspaud) [reply satpy/etc/readers/insat3d_img_l1b_h5.yaml:33]: good question! I'll look into it.

File: satpy/readers/insat3d_img_l1b_h5.py (hunk 4 above)
Comment (mraspaud) [reply satpy/readers/insat3d_img_l1b_h5.py:64]: Good idea!

File: satpy/readers/insat3d_img_l1b_h5.py (hunk 3 above)
Comment (mraspaud) [reply satpy/readers/insat3d_img_l1b_h5.py:57]: indeed, leftover from an earlier version, removing.

File: satpy/readers/insat3d_img_l1b_h5.py (hunk 5 above)
Comment (mraspaud) [reply satpy/readers/insat3d_img_l1b_h5.py:95]: yep!

Comment (mraspaud) [reply satpy/readers/insat3d_img_l1b_h5.py]: Fixed in [be65fb6](https://github.com/pytroll/satpy/pull/2265/commits/be65fb64e234b9ee39ece54cca2daae0aeb0b8a9)

File: satpy/readers/insat3d_img_l1b_h5.py (hunk 4 above)
Comment (mraspaud) [reply satpy/readers/insat3d_img_l1b_h5.py]: fixed in [be65fb6](https://github.com/pytroll/satpy/pull/2265/commits/be65fb64e234b9ee39ece54cca2daae0aeb0b8a9)

File: satpy/readers/insat3d_img_l1b_h5.py (hunk 3 above)
Comment (mraspaud) [reply satpy/readers/insat3d_img_l1b_h5.py]: fiixed in [be65fb6](https://github.com/pytroll/satpy/pull/2265/commits/be65fb64e234b9ee39ece54cca2daae0aeb0b8a9)

File: satpy/etc/readers/insat3d_img_l1b_h5.yaml (hunk 2 above)
Comment (mraspaud) [reply satpy/etc/readers/insat3d_img_l1b_h5.yaml:33]: using the `cfunits.Units` class seems to accept this:
```python
In [15]: 1000 * Units("mW.cm-2.sr-1.micron-1")
Out[15]: <Units: 10000000000 m-1.kg.s-3.rad-2>
```

**Review by djhoese — APPROVED** [review]
Looks good. And I see you have added the conda-forge package to the environment YAML files. I think this means it is available from PyPI and conda-forge @simonrp84 and is treated like a lot of the other readers.

**Comment by djhoese** [comment]
Ah it must complain about "datatree" but the package is "xarray-datatree". Maybe @mraspaud could put a try/except ImportError on the top of the module to give the name of the package instead of the name you import?

**Comment by mraspaud** [comment]
Sure, I'll do that during transfer. Boarding now

**Comment by simonrp84** [comment]
Sorry, I wasn't clear - that's what I meant: The log message says that `datatree` is missing but the actual package is `xarray-datatree`. :)
Try-Except to clarify this sounds good!

**Comment by djhoese** [comment]
Looks good. Merge at your discretion.

**Pull request merged**